### ISSUE TITLE
Runtime Folder Deletes itself if you inject runtime manually to get access to wails runtime

### ISSUE THREAD

**Issue by lckrdg** [wailsapp/wails#2367]
I got the same errors as
```bash
D:\Gitlab\mccl>wails dev
Wails CLI v2.0.0-beta.42 

Executing: go mod tidy -compat=1.17
Executing: wails generate module 
Installing frontend dependencies... 

> mccl@0.0.0 dev
> vite
 

  vite v2.9.15 dev server running at:
 
Vite Server URL: http://localhost:3000/ 
  > Local: http://localhost:3000/ 
  > Network: use `--host` to expose

  ready in 2370ms.

Running frontend DevWatcher command: 'npm run dev' 
Building application for development...
  - Installing frontend dependencies: Done. 
  - Compiling frontend: Done. 
  - Generating bundle assets: Done. 
  - Compiling application:
Build error - open D:\Gitlab\mccl\frontend\wailsjs\runtime\package.json: Access is denied.
No version running, build will be retriggered as soon as changes have been detected
Watching (sub)/directory: D:\Gitlab\mccl 
Using DevServer URL: http://localhost:34115
Using Frontend DevServer URL: http://localhost:3000/
Using reload debounce setting of 100 milliseconds
 
 
To develop in the browser and call your bound Go methods from Javascript, navigate to: http://localhost:34115
[Rebuild triggered] files updated 
  - Generating bundle assets: Done. 
  - Compiling application: Done. 

Added new directory to watcher: D:\Gitlab\mccl\frontend\wailsjs\runtime 
INF | Serving assets from frontend DevServer URL: http://localhost:3000/ 
DEB | WebView2 Runtime Version '104.0.1293.47' installed. Minimum version required: 94.0.992.31.
Error: mkdir D:\Gitlab\mccl\frontend\wailsjs\go\main: Access is denied. 
Development mode exited 
DevWatcher command exited!
DevWatcher command killed!


```
# Situation

Under I monitored, I found some strange things.
1. First, I need to run `wails build` to generate two folders `go` `runtime`.
2. Then, run `wails dev`,  at `  - Installing frontend dependencies: ...` , it will delete the `runtime` folder.
3. Then, wails cannot find the `runtime` and report errors.
 
# Resolve
#### I try to give `frontend\wailsjs\runtime\package.json` the r&w permissions.

1. `wails build` to get the `runtime` folder.
2. Used the `windows file properties` to remove `read`.
3. run `chmod 777 frontend/wailsjs/runtime/package.json`
4. run `wails dev`
5. Unfortunately, `frontend\wailsjs\runtime\package.json` still removed.

#### I try to keep the `runtime` folder.

1. run `wails dev`
2. get the `runtime` `go` folders.
3. run `chmod 777 frontend/wailsjs/runtime/package.json` and use the `windows file properties` to remove `read propertie` of `go` and `runtime` between `Executing: wails generate module ` and `Installing frontend dependencies: ...`
4. Then wails will keep building, nothing reported.
5. Then reported `Error: mkdir frontend\wailsjs\go\main: Access is denied. `
6. I found the `go` folder removed.

#### I try to watch the `go` folder.

1. run `wails dev`
2. get the `go` folders.
3. run `chmod 777 frontend/wailsjs/runtime/package.json` and use the `windows file properties` to remove `read propertie` of `go` and `runtime` between `Executing: wails generate module ` and `Installing frontend dependencies: ...`
4. Then, All errors are gone.

It's amazing, isn't it?

But then I edit files,
```bash
[Rebuild triggered] files updated 
  - Generating bundle assets: Done. 
  - Compiling application:  
Build error - open D:\Gitlab\mccl\frontend\wailsjs\runtime\package.json: Access is denied. 
```
I found `runtime` deleted again.

sad...

So I guess this is a question of permissions.

_Originally posted by @Maicarons2022 in https://github.com/wailsapp/wails/issues/1752#issuecomment-1219007533_

**Comment by taylorlanglopez** [comment]
I have also granted full permissions to everyone just like [this person](https://github.com/wailsapp/wails/issues/1752#issuecomment-1219007533). I have the same issue where the runtime folder just deletes itself if I want to have working code for wails build. 

This is still broken. The runtime folder will always delete itself if you want to inject the runtime into a component for example.

Vue Component (Works but breaks wails dev)
```
<script setup lang="ts">
import * as wailsrt from '../../wailsjs/runtime/runtime'
....
wailsrt.EventsOn("progress", (proginc: string) => {
 //code to update a progress bar
}
...
</script>
```

Vue Component (Doesn't work and throws a compiler error on type-check)
```
<script setup lang="ts">
....
window.runtime.EventsOn("progress", (proginc: string) => {
 //code to update a progress bar
}
...
</script>
```

The second example is "correct" according to the [documentation](https://wails.io/docs/reference/runtime/events), but it just throws a compiler error.

/src/components/File.vue(48,8): error TS2339: Property 'runtime' does not exist on type 'Window & typeof globalThis'

I'm really lost as to how I get access to the runtime inside a component without breaking wails dev. 

My structure is 
src/components
src/views
App.vue
main.ts

I have not modified main.ts (main.js) at all, I added a router to App.vue and il8n.

My App.vue has a <router-view /> from the standard Vue router in order to host the View for each component.

**Comment by chuyijia** [comment]
I got the same error as u，did u work it out?

**Comment by stffabi** [comment]
Thanks for using Wails and bringing up this issue.

Unfortunately AFAIK none of the Core Dev were ever able to reproduce this. I've also followed your reproduction guide and added a vue component with the specified import and the issue didn't occur here.

I suspect this is a timing issue between vite and the creation of the runtime/bindings. Nevertheless I've created a Draft PR which should improve this situation. Would it be possible for you to give PR #2372 a try according to this [guide](https://wails.io/docs/guides/local-development)?

**Comment by chuyijia** [comment]
> Thanks for using Wails and bringing up this issue.
> 
> Unfortunately AFAIK none of the Core Dev were ever able to reproduce this. I've also followed your reproduction guide and added a vue component with the specified import and the issue didn't occur here.
> 
> I suspect this is a timing issue between vite and the creation of the runtime/bindings. Nevertheless I've created a Draft PR which should improve this situation. Would it be possible for you to give PR #2372 a try according to this [guide](https://wails.io/docs/guides/local-development)?

Wow,it worked！！！tks again for the effort u guys made!谢谢~

**Comment by stffabi** [comment]
Awesome, thanks for your feedback and testing it.

Let's see what the other people say that are suffering from this and if it does fix it we could take the PR out of draft.

**Comment by chuyijia** [comment]
> Awesome, thanks for your feedback and testing it.
> 
> Let's see what the other people say that are suffering from this and if it does fix it we could take the PR out of draft.

 after applying the new PR u proposed,the mode of dev works well,but when i altered my code ,the dev watcher retrigerred that error again,which force me to start the  dev mode onemore time

**Comment by stffabi** [comment]
Hmm, I had suspected something like that might happen. I really think there's something that vite is locking there when deleting the files and recreating them. My first impression was that this only happens during startup because starting vite and compiling the app might run in parallel.

We could try to skip rebuilding the runtime, because that won't ever change during reload. Hopefully there's not the same problem with the no normal bindings. Will let you know when the PR is updated.

**Comment by stffabi** [comment]
@chuyijia could you please try the updated version of the PR? Please also make sure to have the correct replace in your `go.mod` in place, since the latest changes also affect the binary itself.

**Comment by chuyijia** [comment]
> @chuyijia could you please try the updated version of the PR? Please also make sure to have the correct replace in your `go.mod` in place, since the latest changes also affect the binary itself.

Gladly,the lose of runtime folder is no concern anymore(your lastest pr worked).Sadly, the go/module folder disappeared as follow 0.0

**Comment by stffabi** [comment]
Seems like Vite (at least on some systems) has problems when a folder is deleted and recreated with new content. Seems like we probably need to sync that folders instead of just deleting and recreating them to work around this.

**Comment by taylorlanglopez** [comment]
Unfortunately, updating my wails is kind of a hassle due to corporate constraints. It takes me multiple days to get stuff updated on the machine where I'm experiencing issues. I can try and recreate my issue with runtime injection and the folder deletion in a public GitHub repo, but that requires me rewriting a significant amount of code. I'll try and get it to this thread by tomorrow. I will also try the PR when I can, thanks for your help.

**Comment by taylorlanglopez** [comment]
This is a pure permissions issue, I haven't tracked down what permission scheme makes the wails dev runtime folder mess itself up, but I cannot replicate the same issue outside of my corporate network. I am going to try the PR you posted today to see if the PR can fix it on my work network.

**Comment by 107295472** [comment]
@taylorlanglopez  Have you solved it, I am also this Access is denied.

**Comment by leaanthony** [comment]
@taylorlanglopez any update?

**Comment by taylorlanglopez** [comment]
> @taylorlanglopez any update?

Not yet. Unfortunately I've been travelling for work and haven't been on that isolated network for awhile. So I haven't been able to look at the permissions structure.

**Comment by mooijtech** [comment]
Also having this issue on Windows, the `wailsjs/runtime` directory is not being created.
`panic: open <path>\wailsjs\runtime\package.json: Access is denied`

The `wailsjs` has `drwxr-xr-x` permissions (755).
Wails CLI version `v2.5.1`.

Works fine on Linux (the `wailsjs` directory was pushed to Git from Linux).

**Comment by haroflow** [comment]
I was also trying to trace these errors, they break the react-ts/angular workflow for me on Windows 10... might happen with all TS templates, not sure.

- `Error: open C:\dev\myproject\frontend\wailsjs\runtime\package.json: Access denied.`
Seems to be caused by vite, as already mentioned. PR 2372 fixes on initialization, but when changing `app.go` I get:
```
 • Generating bindings: 2023/06/25 15:38:22 Error: CreateFile C:\dev\wails-react\frontend\wailsjs\go\main: Acesso negado.
...
 [vite] Internal server error: Failed to resolve import "../wailsjs/go/main/App" from "src\App.tsx". Does the file exist?
```

- `Error: CreateFile C:\dev\myproject\frontend\wailsjs\go\main: Access denied.`
This one happens after I open App.tsx on vscode, seems to be caused by vscode's TS analyzer. Happens on `master` branch too.
---
<details>
<summary>Steps to reproduce both errors on my machine</summary>

> go install github.com/wailsapp/wails/v2/cmd/wails@latest

> wails doctor
```
Wails CLI v2.5.1

 SUCCESS  Done.

# System

OS           | Windows 10 Pro
Version      | 1909 (Build: 18363)
ID           |
Go Version   | go1.20.5
Platform     | windows
Architecture | amd64

# Wails

Version | v2.5.1

# Dependencies

Dependency | Package Name | Status    | Version
WebView2   | N/A          | Installed | 114.0.1823.51
Nodejs     | N/A          | Installed | 16.13.2
npm        | N/A          | Installed | 8.1.2
*upx       | N/A          | Available |
*nsis      | N/A          | Available |
* - Optional Dependency

# Diagnosis

Your system is ready for Wails development!
Optional package(s) installation details:
  - upx : Available at https://upx.github.io/
  - nsis : More info at https://wails.io/docs/guides/windows-installer/

 ♥   If Wails is useful to you or your company, please consider sponsoring the project:
https://github.com/sponsors/leaanthony
```

> wails init -n myproject -t react-ts

> cd .\myproject\

> wails dev

```
...
  • Generating bindings:
2023/06/24 23:53:24 Error: open C:\dev\myproject\frontend\wailsjs\runtime\package.json: Acesso negado.
...
```

> Ctrl+C

> code .

On vscode, open frontend/src/App.tsx.
Go back to powershell

> wails dev

```
Wails CLI v2.5.1

Executing: go mod tidy
  • Generating bindings:
2023/06/24 23:55:01 Error: CreateFile C:\dev\myproject\frontend\wailsjs\go\main: Acesso negado.
...
```
</details>

---

<details>
<summary>Easiest way to test "go\main: Access denied"</summary>

- It happens on `master` branch too.
- Create a react-ts project: `wails init -n myproject -t react-ts`
- Open `myproject` in vscode.
- Open the file App.tsx, this will trigger vscode's TS analyzer. **Without this step the error doesn't happen.**
- On my project's root: `go build -tags bindings -o wailsbindings.exe`
- Run several times at random intervals: `.\wailsbindings.exe`. It fails sometimes, example:
```
PS C:\dev\myproject> go build -tags bindings -o wailsbindings.exe
PS C:\dev\myproject> .\wailsbindings.exe
Error: CreateFile C:\dev\myproject\frontend\wailsjs\go\main: Access denied.
PS C:\dev\myproject> .\wailsbindings.exe
PS C:\dev\myproject> .\wailsbindings.exe
PS C:\dev\myproject> .\wailsbindings.exe
Error: CreateFile C:\dev\myproject\frontend\wailsjs\go\main: Access denied.
PS C:\dev\myproject> .\wailsbindings.exe
PS C:\dev\myproject> .\wailsbindings.exe
PS C:\dev\myproject> .\wailsbindings.exe
PS C:\dev\myproject> .\wailsbindings.exe
Error: CreateFile C:\dev\myproject\frontend\wailsjs\go\main: Access denied.
PS C:\dev\myproject> .\wailsbindings.exe
PS C:\dev\myproject> .\wailsbindings.exe
Error: CreateFile C:\dev\myproject\frontend\wailsjs\go\main: Access denied.
```
</details>

---

More info:
Visual Studio Code, no TS extensions but I'll try disabling the others.
Windows Defender disabled, no AVs installed.

**Comment by leaanthony** [comment]
Thanks for the info @haroflow - Do you think simply overwriting the files would be enough?

**Comment by haroflow** [comment]
@leaanthony You're welcome!

I was thinking about it, not sure if the lock will allow that... don't have much experience with locks on Windows but it's worth a shot.

Some other options, just to throw it out there:
- Move the folder and create another one in its place.
- Don't remove folders, just remove files. Not sure if it's possible and if it would change anything, but I noticed some error was on folder deletion.
- On error, try again a few times to delete/create those files. Sounds bad though, would probably need a few seconds between tries... should be the last resort.
- Stop vite, change files and restart vite? Don't know how long that takes... and doesn't solve vscode's TS analyzer problem.

I also wasn't able to find the exact code that fails yet, will give another try this week.

**Comment by haroflow** [comment]
I spent a few more hours on this:

- Tried to "retry on failure" wherever wails removed/created the bindings just to get a sense of what is going on, and it helped a bit, but it's still failing with: `[vite] Internal server error: Failed to resolve import "../wailsjs/go/main/App" from "src\App.tsx". Does the file exist?`.
  So it seems that Vite tries to read files between wails' changes which is problematic.
  Here are the changes for reference, but it's terrible: https://gist.github.com/haroflow/34f58450b148cf243b5b8d38c54619cd

- Knowing that vite is doing this, I tried a few `vite.config.ts` options:
  - `build: { watch: { buildDelay: 2000 } }`: doesn't work, `wails dev` gets stuck on `Compiling frontend`.
  - `server: { watch: { usePolling: true } }`: **seems to have fixed all errors for me, even with the latest Wails.** Needs more testing.

**Comment by leaanthony** [comment]
Thanks for putting in the time on this 🙏 Sounds promising!

**Comment by id88** [comment]
Wow, I found a solution：
Try to editor `vite.config.ts` options like this
```
export default defineConfig({
  // ...
  server: { watch: { usePolling: true } }
 // ...
})
```
then the error will disappear

**Comment by CLincat** [comment]
I seem to have found a fix step, but I don't know how it works or if it works for anyone else.
I did a test for this:

1. Assuming "main-2" is your application, when you run the `wails dev` command, you get two errors. (white font + red font)
![01](https://github.com/wailsapp/wails/assets/78483995/2ef6baa8-add4-4e58-9ced-b7188b98cc84)


2. To fix this bug, you'll need to create another project. For example `wails init -n test -t svelte`
![02](https://github.com/wailsapp/wails/assets/78483995/7d187f4e-f11c-47ab-bc37-99421bfe3d7f)


3. After that, you can go to the frontend folder and download the npm package you need.
![03](https://github.com/wailsapp/wails/assets/78483995/ef936cac-bbff-4665-b1d3-9e2e84742f7e)


4. Use the command wails dev to run the initial application. This may load something, if this step is missing, it may result in the repair failing. (I don't know how it works)
    We can see an error message in white font, but this does not affect the running.
![04](https://github.com/wailsapp/wails/assets/78483995/6c5703e3-f19c-4590-acf4-be3a98f497ba)


5. Assuming "main-2" is your application, copy the project to the "test" directory. (Except for the `build` and `frontend` folders)
![05](https://github.com/wailsapp/wails/assets/78483995/7010f61a-aeb4-4457-8eb2-b0843202fda5)


6. Delete the frontend/src folder in the test project.
![06](https://github.com/wailsapp/wails/assets/78483995/66e5acd8-3ec1-4fcf-af80-c84b04de3e7c)


7. Similar to step 5, copy the frontend/src folder.
![07](https://github.com/wailsapp/wails/assets/78483995/e500756a-04c2-42da-aaf1-dbff52d656e1)


8. Use the command `wails dev` to run the brand new "test" project.
     We can see an error message in white font, but this does not affect the running.
![08](https://github.com/wailsapp/wails/assets/78483995/c995b2f7-b15e-45e5-bb02-5995e28e6cfd)



During testing, I found that a lot of behaviors lead to fatal errors in red fonts:
    2.1 Using the `npm install` command to get packages while developing a project, and then running the `wails dev` command produces a fatal error with red fonts. If you use `npm install` to get these packages ahead of time, and then develop your project, you won't get an error. ( The question of the order of execution )
    2.2 "main-2" works fine, copy it as "main-3". All files for "main-2" and "main-3" are the same, but the copied "main-3" produces a fatal error with red font when running `wails dev`. ( Problems when copying project folders )


Combining step 4 with testing 2.1, I think it might be an issue with npm package loading?

**Comment by CLincat** [comment]
Also, when encountering a fatal error with red fonts, the command `wails dev` fails to work, but the command `wails build` works fine.

**Comment by leaanthony** [comment]
Are your Dev and non Dev dependencies the same?

**Comment by CLincat** [comment]
Yes, they are the same.

**Comment by CLincat** [comment]
I retested "testing 2.2" and found two things in common.

1. Project "main-3" is working fine, I copied it as "test".
![51](https://github.com/wailsapp/wails/assets/78483995/f5d7655c-3215-4c51-a2f9-45826ea22dc9)


2. Fatal error with red font after using the command `wails dev` in project "test". By contrasting, I found two differences.
![52](https://github.com/wailsapp/wails/assets/78483995/fd7dd873-dfe4-4653-82b8-5e71b5ffe7de)

    2.1  The `frontend/wailsjs/runtime/` folder is deleted.
![53](https://github.com/wailsapp/wails/assets/78483995/4d96fbf9-ff0b-4012-9659-e8d17c3d2a3c)

    2.2 The `frontend/node_modules/.vite/deps/` folder is deleted and replaced by `deps_temp`.
![54](https://github.com/wailsapp/wails/assets/78483995/fe741cd0-46e0-4530-b647-edc5a71e0342)

3. `deps/` and `deps_temp/`.
![55](https://github.com/wailsapp/wails/assets/78483995/285c1361-e0e0-423a-80c0-8054db89c188)
![56](https://github.com/wailsapp/wails/assets/78483995/e3dd0c60-7578-4579-b1f2-9429e0dae11a)



![powershell_CJnneyavGS](https://github.com/wailsapp/wails/assets/78483995/6e6168c6-9ff7-49c1-a7ca-85691b1a40c6)

I found that my previous projects (e.g. "main-2") had the same wrong behavior 2.1 + 2.2
Project main-3 works fine and does not result in a `deps_temp` directory during the `wails dev` process.

**Comment by leaanthony** [comment]
Thanks for the feedback. Noticed that this hasn't had any other feedback since May so closing for now. Feel free to reopen if it's still an issue.

**Comment by YashJainHSR** [comment]
Still getting the same error even with `wails@v2.10.2`

A workaround suggested by @CLincat works but is that a strategic solution?
